### ISSUE TITLE
Make the include setting opt-in, and enable the extension to work globally otherwise

### ISSUE THREAD

**Issue by willdhorn** [jgclark/vscode-todo-highlight#72]
As mentioned in #69, the `include` and `exclude` settings weren't actually wired up previously, so the default behavior was to enable the extension globally. IMO, this should be the default behavior, and user can limit which files types the extension considers by putting filetypes (or languages as mentioned in [this comment](https://github.com/jgclark/vscode-todo-highlight/issues/69#issuecomment-1509329636)) in the `includes` setting[^1], rather than shipping the extension with a predefined list of (mostly web-dev related) file types that it works with and forcing users to add the file types that are specific to how they use vscode.

There are a couple reasons I think this is an improvement:
 - This was the default behavior (unintentional or not) prior to the recent changes
 - it allows users to enable the extension globally without using `**/*`, as the readme explicitly warns against[^2], or, rather absurdly, adding every file extension they are likely to work with
 - As mentioned above, the defaults seem to be aimed primarily at web-dev, but vscode is being used more and more as a general purpose text editor/IDE that it's not unlikely some one programming in python or go installs the extension just to see nothing happen. (this would be made more confusing by #71 and not seeing a setting to modify if they edit their settings with json)

[^1]: It probably makes sense to have a default exclude list to avoid slowing down scanning large files that are not relevant
[^2]: I'm making the assumption that the performance impact mentioned of using `**/*` is different than whatever the extension was doing before the recent changes

**Comment by jgclark** [comment]
Thanks for these good observations and suggestions, which I agree would make for smoother on-boarding. 

The comment about performance impact of `**/*` is inherited from the original author, and AFACIT hasn't been checked. Given they didn't fully test this aspect of the extension, I'm not sure we should now give that too much credence.

As best I can tell, VSCode does use the provided default settings for `include` and `exclude` on a new installation, and it shows them in the visual settings editor. So for exclusions, that is:
"todohighlight.exclude": [
  "**/node_modules/**",
  "**/QuickLook/**",
  "**/dist/**",
  "**/build/**",
  "**/.vscode/**",
  "**/.github/**",
  "**/_output/**",
  "**/*.min.*",
  "**/*.map",
  "**/script.js",
  "**/.next/**"
]

@alkatar21 and @elazarcoh are now helping maintain the extension, and are working on other improvements at the moment. So I'd like their opinion before proceeding.

**Comment by elazarcoh** [comment]
I'm having similar thoughts. 
I also don't like the default value of the include to be a close set of file types (for instance,it doesn't include python and cpp, and would confuse newcomers).
My suggestion is to make special treatment in case of **/*.* (if that really is a performance issue, which I really doubt is relevant to our use case. For me it sounds more relevant in case of matching hundreds of files etc, but I don't really know).

**Comment by alkatar21** [comment]
I am currently very busy, but in a few weeks I have some time. I would also prefer it if `include` would be empty by default and the extension works globally and we possibly extend the exclude list with typical folders you don't want to search (for some common languages).

**Comment by willdhorn** [comment]
When you don't set a value for the setting, it uses the default value set by the extension, which is defined [here](https://github.com/jgclark/vscode-todo-highlight/blob/master/package.json#L227-L248).

However, even when I set the setting to an empty list, a string, or even null, it doesn't highlight anywhere, so there's either an issue with the [`getPaths`](https://github.com/jgclark/vscode-todo-highlight/blob/3d6b43021e324492f0dde20b693f04187ae5d255/src/util.js#L88) function or something else about the way it's handling the setting that's causing it to not evaluate to the `{**/*}`.

On the other hand, I went against the advice or the original author and manually set `"todohighlight.include"` to ` ["**/*.*"]` and have not noticed any kind of performance issue, which makes sense given that was the default behavior before. So if you want to enable the extension to work globally, there doesn't seem to be any problem with doing that like what that warning alludes to.

**Comment by ygoe** [comment]
Yes, I came to the same conclusion now. I want my editor to highlight TODOs, no matter where. And this is a text editor, not something meant to deal with binary data. Opening very large files probably causes performance issues in VSCode before this extension has an effect on it.

**Comment by jgclark** [comment]
I'm releasing 2.1.0 beta here to test these changes to settings:
  - `todohighlight.include` is now dropped.
  - `todohighlight.includedLanguages` is now included as an optional list of languageIds where highlighting will be turned on (e.g. `["typescript", "go"]`). If empty, then all open files will be highlighted. 
  - `todohighlight.exclude` is now called `todohighlight.excludedFiles`. It remains an array of strings of glob pattern that defines files and folders where highlighting will not be turned on.

Test it by getting `vscode-todo-highlight-2.1.0.vsix` and on the extensions sidebar click the ... menu at the top right and select "Install from VSIX...". 
![vsix-install@2x](https://user-images.githubusercontent.com/1671754/233507803-fdd381cf-110a-416e-af54-e52746747aa6.jpg)
You will then need to re-load the VSCode window.

cc @willdhorn @elazarcoh @alkatar21 @ygoe

**Comment by jgclark** [comment]
This is a quick nudge to @willdhorn @elazarcoh @alkatar21 @ygoe hoping that one or more can test this new .vsix (which hopefully contains the updated documenation) so that I can know whether to release it or not.

**Comment by alkatar21** [comment]
I am currently very busy, but I have at least briefly looked over it and left feedback.

**Comment by jgclark** [comment]
I had changed `include` setting to `includedLanguages` in response to the request by @RyanTheTechMan in #69 at https://github.com/jgclark/vscode-todo-highlight/issues/69#issuecomment-1509329636.

In #75 @alkatar21 commented that this stops just selecting a few files or folders, which is true.
So, do I revert this change back to using globs/file paths, but still allowing for the 'all files' case by leaving `include` empty?
Or do I allow both `include` (as above, and essentially no change from v2.0.4), but also allow anything mentioned in `includedLanguages`?

**Comment by alkatar21** [comment]
I would just do include for files for now and `include: []` should then behave exactly as it does now in the `includedLanguages` branch, i.e. allow all files that are not excluded.

I am unsure if `includedLanguages` should be supported for one person, as it is probably rather confusing for most users. If there is more interest in it, you can always support it additionally later. But then we should see if there is something better than `"items": {"type": "string"}` which supports language ids autocompletion.

**Comment by willdhorn** [comment]
I've been using the new .vsix and it's working without having to override anything in my settings. Thanks!

**Comment by ygoe** [comment]
Any update here? Where do I get that beta version? I see that there was no update since then. Does it have the language selection or the file selection now?

**Comment by elazarcoh** [comment]
> Any update here? Where do I get that beta version? I see that there was no update since then. Does it have the language selection or the file selection now?

https://github.com/jgclark/vscode-todo-highlight/blob/jgclark%2Fissue72/vscode-todo-highlight-2.1.0.vsix
